File: analytics-service/src/main/java/com/patientsystem/analyticsservice/service/AnalyticsService.java
package com.patientsystem.analyticsservice.service;

import org.springframework.stereotype.Service;
import java.math.BigDecimal;

import com.patientsystem.analyticsservice.repository.ChargeEventRepository;
import com.patientsystem.analyticsservice.repository.PatientEventRepository;
import java.util.List;
import java.util.Map;
import java.util.LinkedHashMap;
import org.springframework.cache.annotation.Cacheable;

@Service
public class AnalyticsService {
    private final PatientEventRepository patientEventRepository;
    private final ChargeEventRepository chargeEventRepository;
    public AnalyticsService(PatientEventRepository patientEventRepository, ChargeEventRepository chargeEventRepository) {
        this.patientEventRepository = patientEventRepository;
        this.chargeEventRepository = chargeEventRepository;
    }
    @Cacheable("total-active-patients")
    public long getTotalActivePatients(){
        return patientEventRepository.totalCreated() - patientEventRepository.totalDeleted();
    }
    @Cacheable("patient-registrations-per-month")
    public List<Object[]> getPatientRegistrationsPerMonth(int year) {
        return patientEventRepository.countCreatedPerMonth(year);
    }
    @Cacheable("average-age")
    public Double getAverageAge(){
        return patientEventRepository.getAverageAge();
    }
    @Cacheable("annual-revenue")
    public BigDecimal getAnnualRevenue(int year) {
        return chargeEventRepository.getAnnualRevenue(year);
    }
    @Cacheable("revenue-per-category")
    public List<Object[]> getRevenuePerCategory() {
        return chargeEventRepository.getRevenuePerCategory();
    }
    @Cacheable("most-used-treatments")
    public List<Object[]> getMostUsedTreatments() {
        return chargeEventRepository.getMostUsedTreatments();
    }
    @Cacheable("gender-distribution")
    public Map<String, Double> getGenderDistribution() {
        List<Object[]> results = patientEventRepository.getGenderDistribution();
        long total = results.stream().mapToLong(r -> (Long) r[1]).sum();
        Map<String, Double> map = new LinkedHashMap<>();
        for (Object[] row : results) {
            double percentage = (Long) row[1] * 100.0 / total;
            map.put((String) row[0], percentage);
        }
        return map;
    }
}

File: patient-frontend/src/app/services/analytics-service.ts
import { Injectable, Inject } from "@angular/core";
import { HttpClient } from "@angular/common/http";
import { APP_SERVICE_CONFIG, AppConfig } from "../app-config.interface";
import { Observable } from "rxjs";

@Injectable({ providedIn: "root" })
export class AnalyticsService {
  constructor(@Inject(APP_SERVICE_CONFIG) private config: AppConfig, private http: HttpClient) {}

  getActivePatients(): Observable<number> {
    return this.http.get<number>(`${this.config.apiUrl}/api/analytics/active-patients`);
  }

  getAverageAge(): Observable<number | null> {
    return this.http.get<number | null>(`${this.config.apiUrl}/api/analytics/average-age`);
  }

  getGenderDistribution(): Observable<Record<string, number>> {
    return this.http.get<Record<string, number>>(`${this.config.apiUrl}/api/analytics/gender-distribution`);
  }

  getPatientRegistrationsPerMonth(year: number): Observable<any[][]> {
    return this.http.get<any[][]>(`${this.config.apiUrl}/api/analytics/patient-registrations/${year}`);
  }

  getAnnualRevenue(year: number): Observable<string> {
    return this.http.get(`${this.config.apiUrl}/api/analytics/annual-revenue/${year}`, { responseType: 'text' });
  }

  getRevenuePerCategory(): Observable<any[][]> {
    return this.http.get<any[][]>(`${this.config.apiUrl}/api/analytics/revenue-per-category`);
  }

  getMostUsedTreatments(): Observable<any[][]> {
    return this.http.get<any[][]>(`${this.config.apiUrl}/api/analytics/most-used-treatments`);
  }
}
